File: R/svm.R
require(kernlab)
# require(e1071)

x <- read.csv("../dump/dump-EUR_USD.csv", header = FALSE)
names(x) <- c("datetime", "open", "high", "low", "close", "volume")

pip <- 0.0001

x.train <- x[1:400, ]
x.test <- x[401:500, ]
x.n <- NROW(x.test)

fit <- ksvm(close ~ open + high + low, data = x.train, type = "nu-svr", kernel = "vanilladot")
# fit <- svm(close ~ open + high + low, data = x.train, epsilon = 0.1, nu = 0.2, type = "nu-regression", kernel = "linear")
fit

forecast <- as.numeric(predict(fit, x.test))

plot(x.test$close, type = "l")
lines(forecast, col = "red")

mse <- sum(((forecast - x.test$close) / pip) ^ 2) / x.n
mse

diff1 <- round((x.test$open - forecast) / pip, 2)
diff2 <- (x.test$open - x.test$close) / pip
table(diff1 > 0, diff2 > 0)

tail(x.test, 1)
tail(forecast, 1)

File: lib/svm.js
//------------------------------------------------------------------------------
//
// Support Vector Machine analysis
//
//------------------------------------------------------------------------------
"use strict";

var fs = require("fs"),
    svm = require("node-svm");

fs.readFile("../dump/dump.csv", function (err, data) {
    if (err) {
        return;
    }

    processFile(data.toString());
});

function processFile(data) {
    var lines = data.split("\n").slice(0, -1),
        samples = [];

    lines.forEach(function (line) {
        var fields = line.split(","),
            open = parseFloat(fields[1]),
            high = parseFloat(fields[2]),
            low = parseFloat(fields[3]),
            close = parseFloat(fields[4]),
            input = [open, high, low],
            output = close,
            sample = [input, output];

        samples.push(sample);
    });

    runSVM(samples);
}

function runSVM(samples) {
    var trainSamples = samples.slice(0, 400),
        testSamples = samples.slice(400, 500),
        fit = new svm.NuSVR({
            kernelType: "LINEAR",
            reduce: false,
            c: 1,
            nu: 0.2,
            epsilon: 0.1,
            gamma: 0.3333333,
            kFold: 1
        });

    fit
        .train(trainSamples)
        .spread(function () {
            var n = testSamples.length - 1,
                lastInputs = testSamples[n][0],
                lastOutput = testSamples[n][1],
                forecast = fit.predictSync(lastInputs);

            // console.log(model, report);
            console.log(lastInputs, lastOutput, forecast);
    });
}
